doc/taler-bank-transfer.1
.Dd December 14, 2017
.Dt TALER-BANK-TRANSFER 1
.Os
.Sh NAME
.Nm taler-bank-transfer
.Nd
trigger a transfer at the bank
.Sh SYNOPSIS
.Nm
.Op Fl a Ar VALUE | Fl \-amount= Ns Ar VALUE
.Op Fl b Ar URL | Fl \-bank= Ns Ar URL
.Op Fl c Ar FILENAME | Fl \-config= Ns Ar FILENAME
.Op Fl h | \-help
.Op Fl D Ar ACCOUNT | Fl \-debit= Ns Ar ACCOUNT
.Op Fl C Ar ACCOUNT | Fl \-credit= Ns Ar ACCOUNT
.Op Fl s Ar STRING | Fl \-subject= Ns Ar STRING
.Op Fl u Ar USERNAME | Fl \-user= Ns Ar USERNAME
.Op Fl p Ar PASSPHRASE | Fl \-pass= Ns Ar PASSPHRASE
.Op Fl v | \-version
.Sh DESCRIPTION
.Nm
is a command line tool to trigger bank transfers.
.Bl -tag -width Ds
.It Fl a Ar VALUE | Fl \-amount= Ns Ar VALUE
Amount to transfer.
Given in the Taler\-typical format of CURRENCY:VALUE.FRACTION
.It Fl b Ar URL | Fl \-bank= Ns Ar URL
URL at which the bank is operation.
.It Fl c Ar FILENAME | Fl \-config= Ns Ar FILENAME
Use the given configuration file.
.It Fl h | \-help
Print short help on options.
.It Fl D Ar ACCOUNT | Fl \-debit= Ns Ar ACCOUNT
The money should be debited from ACCOUNT.
Specifies the number of the account.
.It Fl C Ar ACCOUNT | Fl \-credit= Ns Ar ACCOUNT
The money should be credited to ACCOUNT.
Specifies the number of the account.
.It Fl s Ar STRING | Fl \-subject= Ns Ar STRING
Use STRING for the wire transfer subject.
.It Fl u Ar USERNAME | Fl \-user= Ns Ar USERNAME
Specifies the username for authentication.
.It Fl p Ar PASSPHRASE | Fl \-pass= Ns Ar PASSPHRASE
Specifies the pass phrase for authentication.
.It Fl v | \-version
Print version information.
.El
.Sh SEE ALSO
.Xr taler-bank-manage 1 ,
.Xr taler.conf 5
.Sh BUGS
Report bugs by using
.Lk https://gnunet.org/bugs/
or by sending electronic mail to
.Aq Mt taler@gnu.org


ChangeLog
Fri 03 May 2019 05:36:10 PM CEST
    Add support for compressing bodies during HTTP uploads to
    exchange and auditor. -CG

Thu 02 May 2019 09:15:37 PM CEST
    Pass hash of denomination public key in operations from
    clients to exchange instead of the (larger) full public key.
    Breaks protocol compatibility, bumping protocol to v3. -CG

Mon Mar 11 03:24:07 CET 2019
	Completed implementation of #5536 (delete private keys once we
	no longer need them). -CG

Sat Mar  2 19:09:43 CET 2019
	Changing denomination key revocation file format and moving them
	to their own directory (preparations for #5536 resolution). -CG

Sat Feb 23 14:47:12 CET 2019
	Actually disable 100 CONTINUE. -CG

Sun Aug 19 15:15:48 CEST 2018
	Increase various limits and rework transaction scopes to
	improve scalability. -CG

Tue Apr  3 23:29:06 CEST 2018
	Releasing Taler v0.5.0

Thu Jan  4 11:55:41 CET 2018
	Fix issue #5234 (aggregator ignoring refunds).
	Misc. minor fixes to the auditor. -CG

Mon Jan  1 23:15:37 CET 2018
	Add TALER_EXCHANGE_refund2() API call to libtalerexchange. -CG

Thu Dec 14 15:32:50 CET 2017
	Replaced taler-exchange-reservemod tool with new taler-bank-transfer
	tool (#5195).  Removed /admin/add/incoming API, replaced by new
	taler-exchange-wirewatch tooling. (#5077).

Sat Dec  9 15:21:50 CET 2071
	Implement refresh protocol optimization, bumping protocol version
	to v2 (#5178). -CG

Thu Nov  2 17:39:40 CET 2017
	Limit amount values to 2^53 as we always wanted (#5167). -CG

Thu Nov  2 16:50:46 CET 2017
	Fixing serious bug in reserve balance calculations, as
	fees were not always properly included in the summation. -CG

Tue Oct 31 13:43:55 CET 2017
	Extended crypto API to provide planchet generation functions
	explicitly (to be used on withdraw, refresh and tipping).
	Renamed a few API calls to better reflect their functions.
	Changed a few function signatures to use new structs. -CG

Wed Oct 18 15:20:29 CEST 2017
	Releasing Taler v0.4.0. -CG

Tue Oct 17 14:32:46 CEST 2017
	Fix building of /keys response to include full auditor
	signature set instead of partial map. -CG

Mon Oct 16 12:10:35 CEST 2017
	Realize JSON-based report formats from auditor (#4963). -CG

Sun Sep 17 16:46:13 CEST 2017
	Implement /keys cherry picking (#4840). -CG

Wed Sep 13 14:06:44 CEST 2017
	Add argument to TALER_EXCHANGE_check_keys_current()
	to force download even if /keys are still considered
	current. -CG

Fri Jul 14 17:38:54 CEST 2017
	Fix #4751 by not allowing denomination deletion (GC)
	to cascade into reserve_out table (and tolerating such
	deletion failures). This way, denominations will be kept
	around until the last reserve that references them (or
	known_coins derived from them) is closed. Naturally, in
	most case we expect reserves to be closed much faster
	than DKs, so in practice this should rarely apply. -CG

Sun Jun 11 17:03:56 CEST 2017
	Finish implementation and testing of automated refunding
	of expired reserves (#4956). -CG

Tue Jun  6 13:53:34 CEST 2017
	Releasing taler-exchange 0.3.0. -CG

Mon Apr 17 01:29:07 CEST 2017
	Add support for HTTP body compression (#4982). -CG

Mon Mar 20 04:37:46 CET 2017
	Implemented first working version of taler-auditor. -CG

Sat Mar 18 03:44:38 CET 2017
	Add 'wire_out' table to exchange DB to track outgoing
	wire transfers. -CG

Fri Nov 18 18:53:30 CET 2016
	Improved error reporting for bogus wire specifications.
	Releasing taler-exchange 0.2.0. -CG

Thu Nov 17 16:37:22 CET 2016
	Added missing idempotency checks for /refresh/reveal. -CG

Thu Nov 17 11:37:56 CET 2016
	Fixed a few cases of missing database rollbacks, causing the
	exchange to be stuck. -CG

Mon Jun 20 17:40:49 CEST 2016
 	Releasing taler-exchange 0.0.2. -CG

Mon Jun 13 16:35:09 CEST 2016
	Splitting HTTP service to run on different listen
	sockets for /admin and /test-APIs and the normal public
	APIs (#4569). -CG

Thu Jun  9 19:17:02 CEST 2016
	Rename /deposit/wtid to /track/transaction and
	/wire/deposits to /track/transfer. -CG

Wed Jun  8 11:32:41 CEST 2016
	Export libtalerfakebank API so we can use it in testcases for the
	merchant as well. -CG

Tue Jun  7 15:13:46 CEST 2016
        Adding public key of the exchange that was used to sign replies
        to applicable callbacks of libtalerexchange. (This will eventually
        be needed by the merchant's backend.) -CG

Mon Jun  6 23:59:05 CEST 2016
	Minor code cleanup, fixing FTBFS when run with just 'make'
	or when header installed to a location not included.
	Releasing taler-exchange 0.0.1. -CG

Wed Jun  1 17:27:36 CEST 2016
	Releasing taler-exchange 0.0.0. -CG


src/bank-lib/taler-bank-transfer.c
/*
  This file is part of TALER
  Copyright (C) 2017 GNUnet e.V.

  TALER is free software; you can redistribute it and/or modify it under the
  terms of the GNU General Public License as published by the Free Software
  Foundation; either version 3, or (at your option) any later version.

  TALER is distributed in the hope that it will be useful, but WITHOUT ANY
  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
  A PARTICULAR PURPOSE.  See the GNU General Public License for more details.

  You should have received a copy of the GNU General Public License along with
  TALER; see the file COPYING.  If not, see <http://www.gnu.org/licenses/>
*/
/**
 * @file taler-bank-transfer.c
 * @brief Execute wire transfer.
 * @author Christian Grothoff
 */
#include "platform.h"
#include <gnunet/gnunet_util_lib.h>
#include <gnunet/gnunet_json_lib.h>
#include <jansson.h>
#include "taler_bank_service.h"

/**
 * Bank URL.
 */
static char *bank_url;

/**
 * Amount to transfer.
 */
static struct TALER_Amount amount;

/**
 * Debit account number.
 */
static unsigned long long debit_account_no;

/**
 * Credit account number.
 */
static unsigned long long credit_account_no;

/**
 * Wire transfer subject.
 */
static char *subject;

/**
 * Username for authentication.
 */
static char *username;

/**
 * Password for authentication.
 */
static char *password;

/**
 * Return value from main().
 */
static int global_ret;

/**
 * Main execution context for the main loop.
 */
static struct GNUNET_CURL_Context *ctx;

/**
 * Handle to access the exchange.
 */
static struct TALER_BANK_AdminAddIncomingHandle *op;

/**
 * Context for running the CURL event loop.
 */
static struct GNUNET_CURL_RescheduleContext *rc;


/**
 * Function run when the test terminates (good or bad).
 * Cleans up our state.
 *
 * @param cls NULL
 */
static void
do_shutdown (void *cls)
{
  if (NULL != op)
  {
    TALER_BANK_admin_add_incoming_cancel (op);
    op = NULL;
  }
  if (NULL != ctx)
  {
    GNUNET_CURL_fini (ctx);
    ctx = NULL;
  }
  if (NULL != rc)
  {
    GNUNET_CURL_gnunet_rc_destroy (rc);
    rc = NULL;
  }
}


/**
 * Function called with the result of the operation.
 *
 * @param cls closure
 * @param http_status HTTP response code, #MHD_HTTP_OK (200) for successful status request
 *                    0 if the bank's reply is bogus (fails to follow the protocol)
 * @param ec detailed error code
 * @param serial_id unique ID of the wire transfer in the bank's records; UINT64_MAX on error
 * @param timestamp timestamp when the transaction got settled at the bank.
 * @param json detailed response from the HTTPD, or NULL if reply was not in JSON
 */
static void
res_cb (void *cls,
        unsigned int http_status,
        enum TALER_ErrorCode ec,
        uint64_t serial_id,
        struct GNUNET_TIME_Absolute timestamp,
        const json_t *json)
{
  op = NULL;
  switch (ec)
  {
  case TALER_EC_NONE:
    global_ret = 0;
    fprintf (stdout,
             "%llu\n",
             (unsigned long long) serial_id);
    break;
  default:
    fprintf (stderr,
             "Operation failed with staus code %u/%u\n",
             (unsigned int) ec,
             http_status);
    if (NULL != json)
      json_dumpf (json,
                  stderr,
                  JSON_INDENT (2));
    break;
  }
  GNUNET_SCHEDULER_shutdown ();
}


/**
 * Main function that will be run.
 *
 * @param cls closure
 * @param args remaining command-line arguments
 * @param cfgfile name of the configuration file used (for saving, can be NULL!)
 * @param cfg configuration
 */
static void
run (void *cls,
     char *const *args,
     const char *cfgfile,
     const struct GNUNET_CONFIGURATION_Handle *cfg)
{
  struct TALER_BANK_AuthenticationData auth;

  ctx = GNUNET_CURL_init (&GNUNET_CURL_gnunet_scheduler_reschedule,
                          &rc);
  GNUNET_assert (NULL != ctx);
  rc = GNUNET_CURL_gnunet_rc_create (ctx);

  auth.method = TALER_BANK_AUTH_BASIC;
  auth.details.basic.username = username;
  auth.details.basic.password = password;
  op = TALER_BANK_admin_add_incoming (ctx,
                                      bank_url,
                                      &auth,
                                      "https://exchange.com/legacy",
                                      subject,
                                      &amount,
                                      debit_account_no,
                                      credit_account_no,
                                      &res_cb,
                                      NULL);
  GNUNET_SCHEDULER_add_shutdown (&do_shutdown,
                                 NULL);
  if (NULL == op)
    GNUNET_SCHEDULER_shutdown ();
}


/**
 * The main function of the reservemod tool
 *
 * @param argc number of arguments from the command line
 * @param argv command line arguments
 * @return 0 ok, 1 on error
 */
int
main (int argc, char *const *argv)
{
  const struct GNUNET_GETOPT_CommandLineOption options[] = {
    GNUNET_GETOPT_option_mandatory
    (TALER_getopt_get_amount ('a',
                              "amount",
                              "VALUE",
                              "value to transfer",
                              &amount)),
    GNUNET_GETOPT_option_mandatory
    (GNUNET_GETOPT_option_string ('b',
                                  "bank",
                                  "URL",
                                  "base URL of the bank",
                                  &bank_url)),
    GNUNET_GETOPT_option_help ("Deposit funds into a Taler reserve"),
    GNUNET_GETOPT_option_mandatory
    (GNUNET_GETOPT_option_ulong ('C',
                                 "credit",
                                 "ACCOUNT",
                                 "number of the bank account to credit",
                                 &credit_account_no)),
    GNUNET_GETOPT_option_mandatory
    (GNUNET_GETOPT_option_ulong ('D',
                                 "debit",
                                 "ACCOUNT",
                                 "number of the bank account to debit",
                                 &debit_account_no)),
    GNUNET_GETOPT_option_mandatory
    (GNUNET_GETOPT_option_string ('s',
                                  "subject",
                                  "STRING",
                                  "specifies the wire transfer subject",
                                  &subject)),
    GNUNET_GETOPT_option_mandatory
    (GNUNET_GETOPT_option_string ('u',
                                  "user",
                                  "USERNAME",
                                  "username to use for authentication",
                                  &username)),
    GNUNET_GETOPT_option_mandatory
    (GNUNET_GETOPT_option_string ('p',
                                  "pass",
                                  "PASSPHRASE",
                                  "passphrase to use for authentication",
                                  &password)),
    GNUNET_GETOPT_OPTION_END
  };

  GNUNET_assert (GNUNET_OK ==
                 GNUNET_log_setup ("taler-bank-transfer",
                                   "WARNING",
                                   NULL));
  global_ret = 1;
  if (GNUNET_OK !=
      GNUNET_PROGRAM_run (argc, argv,
                          "taler-bank-transfer",
			  "Execute bank transfer",
			  options,
			  &run, NULL))
    return 1;
  return global_ret;
}

/* end taler-bank-transfer.c */
